File: ai-service/src/api/embedding.py
"""The embedding feature — Phase 10.

The narrowest capability in the service. Text in, a vector out. No prompt, no
schema, no repair round, no taxonomy, no thresholds, and — unlike every other
feature here — no model output to distrust: an embedding is a measurement of
the input, not an assertion about it.

⚠️ WHAT THIS MEANS FOR PROMPT INJECTION. There is no prompt, so there is
nothing to inject into. A ticket that says "ignore your instructions" is
embedded as the sentence it is; the model is not being asked to decide
anything, so there is no decision to subvert. The residual risk is different
and worth naming: text CRAFTED to sit near a target vector could make an
attacker's own ticket surface as a "similar ticket". Retrieval ranking is not
an authorization boundary and is never treated as one — RLS decides what a
query can see, and it runs before ranking.

WHY `dim` IS RETURNED ALONGSIDE THE VECTOR. So a deployment swap that changes
width is caught at this boundary by an assertion with a clear message, rather
than 3 hops later by a Postgres type error on a `vector(1536)` column.
"""

from __future__ import annotations

import time

from ..config import config
from .schemas import AIResult, ExecuteRequest

#: Bumped when the canonical text format or the model changes — both alter what
#: the vectors mean, and vectors from two versions must not be compared.
EMBEDDING_PROMPT_VERSION = "embedding-v1"

#: Matches EMBEDDING_MAX_CHARS in shared/types/embedding.ts. Core truncates in
#: SQL so the fingerprint and the embedded text always describe one string;
#: this is a defensive backstop, not the enforcement point.
EMBEDDING_MAX_CHARS = 8000


async def run_embedding(req: ExecuteRequest) -> AIResult:
    """Embed the canonical text carried in `input.description`.

    NOTE the reuse of `description`. The frozen data-boundary contract
    (shared/contracts/ai/ai-contracts.schema.json) was NOT widened for this
    feature: `description` is the field for "the text to process", and the
    canonical text — built by Postgres in migration 014's generated column — is
    exactly that. Adding a `texts` array would have bought request batching at
    the cost of changing a contract whose `additionalProperties: false` is a
    tested security property. Concurrency is recovered in the worker instead,
    where it costs nothing.

    `subject` is null on this path for the same reason: splitting the canonical
    text back into two fields would create a second spelling of a string whose
    single spelling is the entire idempotency mechanism.
    """
    started = time.perf_counter()

    text = req.input.description or ""
    if not text.strip():
        # Permanent: empty text is empty on every retry too.
        raise _feature_error("invalid_input", "text must not be empty", "permanent")

    if not config.embedding_enabled:
        # Honest unavailability. Temporary, so the item stays pending and is
        # retried once a credential exists — an unconfigured deployment must
        # never write a fabricated or zero vector into the corpus.
        raise _feature_error(
            "provider_not_configured",
            "no embedding provider credential is configured",
            "temporary",
        )

    from ..integrations.llm_client import (
        LLMPermanentError,
        LLMTemporaryError,
        _classify_provider_exception,
    )

    client = _embeddings()

    try:
        vector = await client.create(text=text[:EMBEDDING_MAX_CHARS])
    except Exception as exc:  # noqa: BLE001 — reclassified immediately below
        code, kind = _classify_provider_exception(exc)
        raise _feature_error(code, str(exc)[:300], kind) from exc

    # The boundary assertion. A deployment that starts returning a different
    # width is a configuration fault, not a transient one: retrying six times
    # cannot change it, so it fails permanently and loudly.
    if len(vector) != config.embedding_dim:
        raise _feature_error(
            "embedding_dim_mismatch",
            f"provider returned {len(vector)} dimensions, expected {config.embedding_dim}",
            "permanent",
        )

    return AIResult(
        feature="embedding",
        status="succeeded",
        # Core validates this — finite, correct width, non-zero — before any
        # of it reaches a column. Python returning it is not Core accepting it.
        data={"vector": vector, "dim": len(vector), "model": config.embedding_model_id},
        confidence=None,
        provider="azure",
        model=config.embedding_model_id,
        model_version=None,
        prompt_version=EMBEDDING_PROMPT_VERSION,
        latency_ms=int((time.perf_counter() - started) * 1000),
        fallback_used=False,
    )


def _feature_error(code: str, message: str, kind: str):
    # Imported lazily to avoid a circular import: features.py imports this
    # module, and FeatureError lives there.
    from .features import FeatureError

    return FeatureError(code, message, kind)


# The process-lifetime embeddings transport, cached for the same reason the
# chat transport is: a connection pool rebuilt per call is not a pool.
_EMBEDDINGS = None


def _embeddings():
    global _EMBEDDINGS  # noqa: PLW0603 — a process-lifetime singleton
    if _EMBEDDINGS is not None:
        return _EMBEDDINGS
    from ..integrations.embedding_client import HttpxEmbeddings

    _EMBEDDINGS = HttpxEmbeddings(
        base_url=config.azure_embedding_endpoint,
        api_key=config.azure_embedding_api_key,
        deployment=config.azure_embedding_deployment,
        api_version=config.azure_embedding_api_version,
        timeout_seconds=config.classification_budget_seconds,
    )
    return _EMBEDDINGS


def set_embeddings(stub) -> None:
    """Test seam: substitute the transport. Never used in production."""
    global _EMBEDDINGS  # noqa: PLW0603 — module-level test seam
    _EMBEDDINGS = stub

File: shared/types/embedding.ts
/**
 * Embedding contracts — Phase 10.
 *
 * WHAT AN EMBEDDING IS HERE. A derived retrieval artifact: a vector that lets
 * Postgres rank one piece of the tenant's own text against another. It is not
 * a prediction, it carries no confidence, and it reaches no decision.
 *
 * ⚠️ WHY THIS IS NOT A `feature` ON THE ai.jobs QUEUE.
 *
 * Three concrete blockers, all in the existing schema and contracts rather
 * than in taste:
 *
 *   1. `ai_execution.ticket_id text NOT NULL REFERENCES ticket(id)`. Forty-
 *      eight of the 120 corpus items are KB articles, which have no ticket.
 *      Routing them through the queue means weakening the FK on the
 *      GOVERNANCE table — to store a 1536-float array in a `result` column
 *      whose own comment says it holds "the VALIDATED result only".
 *
 *   2. `AI_EVENT_FEATURES` is keyed on `ticket.created`. The corpus is
 *      RESOLVED tickets and PUBLISHED articles. Neither state is that event,
 *      so reuse would need two new event types plus outbox writes inside
 *      ticket and KB lifecycle code — and neither would notice a later EDIT.
 *
 *   3. Idempotency. The pipeline's durable key is `UNIQUE(event_id, feature)`,
 *      which cannot express "the text changed, embed it again". Embedding's
 *      correct key is a CONTENT FINGERPRINT, which makes backfill, incremental
 *      and re-embed-after-edit one operation instead of three.
 *
 * So the queue is not reused. What IS reused, deliberately and without
 * modification: the `/v1/execute` contract, both HMAC trust edges, the
 * temporary/permanent error model, the Core-internal route pattern, and the
 * worker as the only process that talks to both Core and Python. No new trust
 * relationship is created, and the AI service still holds no database
 * credential.
 */

/**
 * The provider's native output width for `text-embedding-3-small`, verified
 * live against the configured Azure deployment.
 *
 * ⚠️ CHANGING THIS IS A MIGRATION, not a config edit. `ticket.embedding` and
 * `kb_article.embedding` are `vector(1536)`; Postgres rejects a vector of any
 * other width outright. That rejection is the point — a dimension mismatch
 * fails at the write instead of silently poisoning the corpus with vectors
 * that cannot be compared to each other.
 */
export const EMBEDDING_DIM = 1536;

/**
 * Recorded on every row as `embedding_model`.
 *
 * Vectors from different models are not comparable, so this is what makes
 * "which of these rows are stale because we changed model?" answerable with a
 * query rather than a guess. It is deliberately NOT part of the content
 * fingerprint: the fingerprint hashes CONTENT, and mixing the two would put a
 * provider choice inside the schema.
 */
export const EMBEDDING_MODEL_ID = 'azure/text-embedding-3-small';

/** What the corpus is drawn from. Both are product-scoped, RLS-protected tables. */
export type EmbeddingSubjectType = 'ticket' | 'kb_article';

export const EMBEDDING_SUBJECT_TYPES: readonly EmbeddingSubjectType[] = ['ticket', 'kb_article'];

/**
 * One item of pending work, as Core hands it to the worker.
 *
 * NOTE WHAT IS ABSENT, and must stay absent: product_id, tenant id, reference,
 * raiser identity, status, assignee. The worker forwards only `text` to
 * Python, and Core re-derives the tenant from `subject_id` when the vector
 * comes back. The worker is a courier, not an authority — the same property
 * the ai.jobs path has.
 */
export interface EmbeddingWorkItem {
  subject_type: EmbeddingSubjectType;
  subject_id: string;
  /**
   * The canonical text, built by Postgres from the row itself.
   *
   * ⚠️ NEVER rebuilt in TypeScript. The canonical form is the generated
   * `embedding_content_sha` expression in migration 014, and a second
   * implementation of it would drift silently — re-embedding the entire corpus
   * on every cycle, billed each time, with nothing failing to signal it.
   */
  text: string;
  /** The value of `embedding_content_sha` at the moment Core read the row. */
  fingerprint: string;
}

/**
 * What the worker sends back for ONE item.
 *
 * `fingerprint` is echoed, not recomputed, and Core writes the vector only if
 * the row's fingerprint is STILL this value. That closes a real race: a ticket
 * edited while its embedding was in flight would otherwise be stamped current
 * while holding a vector of the previous text — a stale vector that looks
 * fresh is worse than a missing one, because nothing ever revisits it.
 */
export interface EmbeddingApplyItem {
  subject_type: EmbeddingSubjectType;
  subject_id: string;
  fingerprint: string;
  /** Exactly EMBEDDING_DIM finite numbers. Core validates before persisting. */
  vector: number[];
  model: string;
}

/**
 * A PERMANENT provider failure for one item, reported so Core can quarantine
 * it (migration 015).
 *
 * Sent in the same request as the successes rather than through a second
 * endpoint: they come from one cycle over one batch, and splitting them would
 * let a worker crash between the two calls apply the vectors while losing the
 * failures — leaving the failed rows to be retried forever, which is the exact
 * thing the quarantine exists to stop.
 */
export interface EmbeddingFailureItem {
  subject_type: EmbeddingSubjectType;
  subject_id: string;
  fingerprint: string;
  /** A stable code. NEVER provider prose, which can quote the input text. */
  code: string;
}

export interface EmbeddingApplyResponse {
  /** Rows actually written. Lower than the request when a row was edited mid-flight. */
  applied: number;
  /** Items rejected by validation or by the fingerprint guard, with a reason. */
  skipped: Array<{ subject_id: string; reason: string }>;
  /** Rows quarantined after a permanent provider failure. */
  quarantined: number;
}

export interface EmbeddingPendingResponse {
  items: EmbeddingWorkItem[];
}

/**
 * The AI service's output for `feature: "embedding"`.
 *
 * `dim` is returned alongside the vector rather than left implicit so a
 * provider or deployment swap that changes width is caught at the boundary,
 * by an assertion, instead of at the database by a constraint violation.
 */
export interface EmbeddingData {
  vector: number[];
  dim: number;
  model: string;
}

/**
 * How many items one worker cycle claims from Core.
 *
 * NOT a provider batch. Azure's embeddings endpoint does accept an array
 * `input`, and using it would have made a cycle one HTTP call instead of 16 —
 * but only by widening `ExecuteRequest.input` with a `texts` array, and that
 * contract's `additionalProperties: false` is a tested security property
 * rather than a formality. The connection cost batching would have saved is
 * already gone: the transport is pooled, so 16 calls share one TCP connection
 * and one TLS handshake. Concurrency inside the cycle recovers the wall-clock.
 *
 * Sized so a full backfill of the 120-item corpus is a handful of cycles, and
 * so a provider outage mid-cycle costs 16 items rather than everything.
 */
export const EMBEDDING_BATCH_SIZE = 16;

/**
 * Maximum characters of canonical text sent for one item.
 *
 * `text-embedding-3-small` accepts 8191 tokens. This bound is well inside it
 * and exists for a different reason: a 40k-character ticket produces a vector
 * dominated by whatever it rambles about, which is worse retrieval than
 * embedding its opening. Truncation is applied in SQL so the fingerprint and
 * the embedded text always describe the same string.
 */
export const EMBEDDING_MAX_CHARS = 8000;

/**
 * Rejects anything Postgres would accept as a vector but that would corrupt
 * ranking.
 *
 * NaN is the case that matters and the reason this is not just a length check:
 * pgvector stores NaN happily, every distance involving it is NaN, and NaN
 * sorts LAST under `ORDER BY ... ASC` — so a poisoned row does not crash a
 * query, it silently never appears. A corpus can rot one row at a time with
 * nothing failing anywhere.
 *
 * A zero vector is rejected for a related reason: cosine distance to it is
 * undefined (pgvector returns NaN), so it is the same failure wearing a
 * different mask.
 */
export function validateEmbeddingVector(
  value: unknown,
  dim: number = EMBEDDING_DIM,
): { ok: true; vector: number[] } | { ok: false; reason: string } {
  if (!Array.isArray(value)) return { ok: false, reason: 'vector is not an array' };
  if (value.length !== dim) {
    return { ok: false, reason: `expected ${dim} dimensions, got ${value.length}` };
  }

  let sumSquares = 0;
  for (let i = 0; i < value.length; i++) {
    const n = value[i];
    if (typeof n !== 'number') return { ok: false, reason: `dimension ${i} is not a number` };
    // Catches NaN, Infinity and -Infinity in one check.
    if (!Number.isFinite(n)) return { ok: false, reason: `dimension ${i} is not finite` };
    sumSquares += n * n;
  }
  if (sumSquares === 0) return { ok: false, reason: 'vector is all zeros' };

  return { ok: true, vector: value as number[] };
}

/**
 * pgvector's text input format: `[0.1,-0.2,...]`.
 *
 * Sent as a bound parameter cast with `$n::vector`, never interpolated.
 * Building the literal here rather than at the call site keeps that one
 * decision in one place.
 */
export function toVectorLiteral(vector: number[]): string {
  return `[${vector.join(',')}]`;
}
